### DISCUSSION TITLE
Pico: host a memory device (SDcard via SPI)

### DISCUSSION THREAD

**Discussion by OlafM2015** [micropython/micropython#9241] [Ideas]
We are building a Music player for people with dementia using the Pico and the `vs1053.py` module. The Pico reads a SDcard via SPi and sends this to an mp3 decoder  (VS1053). 
A great feature would be, that the caretaker can update the music on the SDcard via the micoUSB on the Pico, such that the people with dementia can have a personalized music list.
For this features, we would need  that the Pico is in host mode for the SDcard.  Sofar, I was not able to find a solution. 
What would you recommend?

**Comment by scruss** [comment]
Can't comment on being able to access an SD Card in the way you hope to do, but have you been in contact with [Playlist for Life – Personal music for dementia](https://www.playlistforlife.org.uk/)? They're a charity who provide support for people experiencing dementia through music rehab. Good and useful devices to do this are hard to find. The [Memory Loss Music Player](https://makersmakingchange.com/project/dementia-friendly-music-player/) is a device to help with this, hosted on my former employer's site, [Makers Making Change](https://makersmakingchange.com/)

**Comment by OlafM2015** [comment]
Thanks for the link https://www.playlistforlife.org.uk/, looks really helpful, as making playlists is not so easy.
A musicalplayer is difficult as the user testing is complex and takes considerable time. 
We are a Dutch foundation and developed the Oradio in the last 3 years by many usertest. See www.StichtingOradio.nl ( sorry for the Dutch) and www.Oradio.tech were we would like an open source platform such that many other can build such a musicplayer. That is also the reason why we switch from a Raspberry pi to a Pico which is much better available.
Thanks for the links good references.

**Comment by jimmo** [comment]
> For this features, we would need that the Pico is in host mode for the SDcard.

I'm not quite sure what you mean by this?

I would imagine that what you want is for the Pico to essentially act as a SD Card Reader (i.e. if you plug your device into a PC, it looks like a USB Flash Drive and you can copy .mp3 files to/from the device).

In which case you want the Pico to be in USB device mode, not USB host mode, which we do actually support but on the Pico only for accessing the internal Flash. What you're describing would work on STM32 though, in boot.py you can set up the MSC device such that it uses the SDCard. We plan to add this to the rp2 (Pico) port eventually by allowing the USB MSC driver to be configured with any arbitrary range of a block device. Can't say when that will be ready though unfortunately.

**Reply by OlafM2015** [reply]
Indeed, we would like to acces the SDcard via the microUSB connection on the Rpi Pico. Which would enable users easily to update the music on the Sd card.

**Comment by peterhinch** [comment]
To clarify, this is a continuation of [this discussion](https://github.com/peterhinch/micropython-vs1053/issues/3). I am unclear about the exact requirement. I can see two options:
 1. Use USB host mode to mount a memory stick or other USB memory device.
 2. Use an SPI interface to connect an SD card socket.

Option 2 is easy and I have tested in this mode.

I recommended that @OlafM2015 posted here because I am unsure whether there is any scope for supporting option 1.

**Reply by OlafM2015** [reply]
To make it easy for the caretakers, the media to have the Mp3s on, in order of preference:
1) an micro SD card, which is inside the music player, which can be read/write via the micro USB connection on the Pico
3) USB stick, which is plugged in the rear of the music player
3) an micro SD card, which is removable and can be updated on PC/laptop
Option 1) is favourable because of the robust approach, no lose parts , low mechanical impact
Option 3) is the current approach, which is cumbersome and unpractical for caretakers (small micro SD card)
See also attached pictures

For reference, the music player HW consists of Rpi Pico, VS1053 mp3 decoder, MAX98357 I2S amplifiers, Touch interface. With the help of peterhinch VS1053.py class, we have now a robust working Music player running on the Pico. 

It would be great, if option 1 or 2 would be enabled.
Hope this clarifies the question and requirements. Many thanks for suggestions and hints.
<img width="1166" alt="Screenshot 2022-09-08 at 18 44 49" src="https://user-images.githubusercontent.com/11895505/189183175-c4ace529-400c-40fd-bd8a-de5eee55b848.png">
<img width="1173" alt="Screenshot 2022-09-08 at 18 49 35" src="https://user-images.githubusercontent.com/11895505/189183198-84267357-7633-4622-b49f-42dc17db5d15.png">

**Comment by peterhinch** [comment]
OK, now I understand.

As far as I know option 2 (USB host mode) is not supported, but I was hoping for guidance from someone more sure of the facts.

Option 1 may be possible using `mpremote`. The micro SD card can be mounted in the Pico's filesystem: `mpremote` could run the script that does this. It could then copy the files to the card. I guess you'd need to write a simple PC application to make this process user friendly.

Or maybe someone can think of a better way as this is rather outside my experience. Would an existing IDE help in managing a remote filesystem?
